src/algebra/euclid-algorithm.md

```diff
@@ -0,0 +1,119 @@
+	return b ? gcd (b, a % b) : a;
+}
+```
+
+Finally, we present a non-recursive form of implementation of the algorithm:
+
+```cpp
+int gcd (int a, int b) {
+	while (b) {
+ a %= b;
+ swap (a, b);
+	}
+	return a;
+}
+```
+
+## The Proof of correctness
+
+First, observe that at each successive iteration of the Euclidean algorithm, it's second argument strictly decreases, therefore, since it is always non-negative, the algorithm must terminate.
+
+For the proof of correctness, we need to show $gcd(a, b) = gcd(b, a\mod b)$ for all $a \geq 0$, $b > 0$.
+
+We will show that the quantity on the left side of the equation, can be written as the value on the left. Obviously, this would mean that the left and right sides are the same and prove Euclid's algorithm.
+
+Let $d = gcd(a, b)$. Then, by definition $d\mid a$ and $d\mid b$.
+
+New writing the remainder on dividing $a$ by $b$:
+
+![image](&imgroot&/euclid3.png)
+
+But then, it follows:
+
+![image](&imgroot&/euclid4.png)
+
+So, remembering the statement $d\mid b$, we obtain the system:
+
+![image](&imgroot&/euclid5.png)
+
+We now use the following fact: if for any three numbers $p$, $q$, and $r$, if $p\mid q$ and $p\mid r$ then $p\mid gcd(q, r)$, In our case, we get:
+
+![image](&imgroot&/euclid6.png)
+
+Or, by substituting $d$ by it's definition($gcd(a, b)$) we get:
+
+![image](&imgroot&/euclid7.png)
+
+So, we have shown that the left side divides the right. The second half of the proof is similar.
+
+## Time complexity
+
+The running time of the algorithm is estimated by Lame's theorem, which establishes a surprising connection between the euclidean algorithm and the Fibonacci sequence:
+
+If $a > b \geq 1$ and $b < F_n$ for some $n$, the euclidean algorithm performs no more than $n-2$ recursive calls.
+
+Moreover, one can show that the upper bound of this theorem is optimal. When $a = F_n$ and $b = F_{n-1}$, $gcd(a, b)$ performs $n-2$ recursive calls. In other words, consecutive fibonacci numbers are the worst case input for Euclid's algorithm.
+
+Given that Fibonacci numbers grow exponentially(as a constant with respect to $n$), we find that the Euclidean algorithm is performed in $O(\log min(a, b))$
+
+## LCM(Least common multiple)
+
+Calculating least common multiple (least common multiplier, lcm) is reduced to calculating $gcd$ with the following simple statement:
+
+![image](&imgroot&/euclid8.png)
+
+Thus, the calculation of the LCM can also be done using the euclidean algorithm with the same time complexity
+
+```cpp
+int lcm (int a, int b) {
+	return a / gcd (a, b) * b;
+}
+```
+(here it is advantageous to first divide $gcd$ and then multiply by $b$, as it helps avoid overflow in some cases)
+
+##Literature
+
+- Thomas Cormen, Charles Leiserson, Ronald Rivest, Clifford Stein. Algorithms: Construction and analysis [2005]
+
+## Problems
+
+- [GCD and LCM[easy]](https://www.codechef.com/problems/FLOW016)
+
+####Note: Translated by [NibNalin](http://codeforces.com/profile/NibNalin)
```

Comment:
Per https://e-maxx-eng.appspot.com/contrib.html, we've agreed to track authorship of pages not with sole author but using Github edits history. Let's remove this line. You can add link to your CF profile in your Github profile if you want to preserve your CF persona.